x/auctions/keeper/hooks.go

```diff
@@ -17,50 +18,53 @@ func (k Keeper) BeforeEpochStart(_ sdk.Context, _ string, _ int64) {
 
 // AfterEpochEnd ends the current auction and distributes the rewards to the winner
 func (k Keeper) AfterEpochEnd(ctx sdk.Context, epochIdentifier string, _ int64) {
+	// Create a cached context that is committed only
+	// if not errors happen.
+	ctxCache, writeFn := ctx.CacheContext()
+
 	// If it's not the weekly epoch, no-op
 	if epochIdentifier != epochstypes.WeekEpochID {
 		return
 	}
 
-	params := k.GetParams(ctx)
+	params := k.GetParams(ctxCache)
 	if !params.EnableAuction {
 		return
 	}
 
-	lastBid := k.GetHighestBid(ctx)
+	lastBid := k.GetHighestBid(ctxCache)
 
 	// Distribute the awards from the last auction
-	if isValidBid(lastBid) {
-		bidWinner, err := sdk.AccAddressFromBech32(lastBid.Sender)
-		if err != nil {
-			return
-		}
 
+	// lastBid.Sender can be "", "invalidBech32" or "validBech32".
+	bidWinner, err := sdk.AccAddressFromBech32(lastBid.Sender)
+
+	// A valid bid is one in which lastBid.Sender is "validBech32" and the
+	// bid.Amount.Amount is positvie.
+	if err == nil && lastBid.Amount.Amount.IsPositive() {
 		moduleAddress := k.accountKeeper.GetModuleAddress(types.ModuleName)
-		coins := k.bankKeeper.GetAllBalances(ctx, moduleAddress)
+		coins := k.bankKeeper.GetAllBalances(ctxCache, moduleAddress)
 
-		remainingCoins := sdk.NewCoins()
-		for _, coin := range coins {
-			if coin.Denom != utils.BaseDenom {
-				remainingCoins = remainingCoins.Add(coin)
-			}
-		}
+		remainingCoins := removeBaseCoinFromCoins(coins)
 
-		// Burn the Evmos Coins from the module account
-		if err := k.bankKeeper.BurnCoins(ctx, types.ModuleName, sdk.NewCoins(lastBid.Amount)); err != nil {
+		// Burn the Evmos Coins from the module account.
+		if err := k.bankKeeper.BurnCoins(ctxCache, types.ModuleName, sdk.NewCoins(lastBid.Amount)); err != nil {
 			return
 		}
 
-		// Send the remaining Coins from the module account to the auction winner
-		if err := k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, bidWinner, remainingCoins); err != nil {
+		// Send the remaining Coins from the module account to the auction winner.
+		fmt.Println("Try to send")
+		fmt.Println(len(remainingCoins))
+		if err := k.bankKeeper.SendCoinsFromModuleToAccount(ctxCache, types.ModuleName, bidWinner, remainingCoins); err != nil {
+			fmt.Println("Error in sending coins")
 			return
 		}
 
 		// Clear up the bid in the store
-		k.deleteBid(ctx)
+		k.deleteBid(ctxCache)
```

Comment:
the ctxCache seems to only be useful within the `if err == nil && lastBid.Amount.Amount.IsPositive() {` block
```suggestion
		k.deleteBid(ctxCache)
		writeFn()
```

So we can just create the cache ctx only when that condition is met. The usage if the ctx before this block is just for querying